File: week1_baseline/ruby/09_global_executable/lib/boukensha_loader.rb
# BoukenshaLoader resolves which step folder and config directory to use, then
# boots the REPL.
#
# Each setting is resolved independently in this order:
#   1. BOUKENSHA_PATH / BOUKENSHA_DIR environment variable
#   2. boukensha_path / boukensha_dir in ~/.boukensharc
#   3. The bundled lib / ~/.boukensha default
#
# Examples:
#   boukensha                                                              # uses bundled lib + ~/.boukensha
#   BOUKENSHA_PATH=~/Sites/boukensha/04_api_client boukensha              # loads step 4
#   BOUKENSHA_DIR=~/projects/mybot/.boukensha boukensha                   # custom config dir
require "yaml"
require "rubygems"
require_relative "boukensha/version"

module BoukenshaLoader
  # Absolute path to this gem's own bundled boukensha lib.
  BUNDLED_LIB = File.expand_path("../boukensha.rb", __FILE__)

  def self.rc_file
    File.expand_path("~/.boukensharc")
  end

  def self.load_rc
    return {} unless File.exist?(rc_file)

    parsed = YAML.safe_load(
      File.read(rc_file),
      permitted_classes: [],
      aliases: false
    )

    case parsed
    when Hash
      parsed
    when String
      # Backward compatibility with the original single-path format.
      { "boukensha_path" => parsed }
    when nil
      {}
    else
      abort "boukensha: #{rc_file} must contain a YAML mapping"
    end
  rescue Psych::SyntaxError => e
    abort "boukensha: invalid YAML in #{rc_file}: #{e.message}"
  end

  def self.expand_rc_path(path)
    return nil unless path.is_a?(String)
    return nil if path.strip.empty?

    File.expand_path(path, File.dirname(rc_file))
  end

  def self.resolve
    rc = load_rc

    # Apply this before requiring the selected implementation. An explicit
    # environment variable always wins over the rc file.
    rc_config_dir = expand_rc_path(rc["boukensha_dir"])
    ENV["BOUKENSHA_DIR"] = rc_config_dir if !ENV["BOUKENSHA_DIR"] && rc_config_dir

    source = ENV["BOUKENSHA_PATH"] || expand_rc_path(rc["boukensha_path"])
    return BUNDLED_LIB unless source

    dir = File.expand_path(source)
    main = File.join(dir, "lib", "boukensha.rb")
    return main if File.exist?(main)

    abort <<~MSG
      boukensha: no lib/boukensha.rb found at:
             #{dir}
             Check BOUKENSHA_PATH or #{rc_file}.
    MSG
  end

  def self.load_and_start_repl
    main = resolve
    step_dir = File.dirname(File.dirname(main))

    puts "[boukensha] loading from: #{step_dir}" if ENV["BOUKENSHA_DEBUG"]

    require main

    unless Boukensha.respond_to?(:repl)
      abort <<~MSG
        boukensha: the step at #{step_dir}
               does not support the interactive REPL (added in step 7).
               Run its examples directly, e.g.:
                 ruby #{step_dir}/examples/*.rb
               Or point BOUKENSHA_PATH at step 7 or later.
      MSG
    end

    Boukensha.repl
  end

  def self.version(io: $stdout)
    io.puts "boukensha #{Boukensha::VERSION}"
  end

  def self.doctor(io: $stdout)
    rc = load_rc
    implementation = resolve
    config_dir = ENV["BOUKENSHA_DIR"] ||
      expand_rc_path(rc["boukensha_dir"]) ||
      File.expand_path("~/.boukensha")
    settings_file = File.join(config_dir, "settings.yaml")
    settings = if File.exist?(settings_file)
                 YAML.safe_load(
                   File.read(settings_file),
                   permitted_classes: [],
                   aliases: false
                 ) || {}
               else
                 {}
               end
    player = settings.fetch("tasks", {}).fetch("player", {})
    provider = player["provider"]
    model = player["model"]
    credential_name = {
      "anthropic" => "ANTHROPIC_API_KEY",
      "openai" => "OPENAI_API_KEY",
      "gemini" => "GEMINI_API_KEY",
      "ollama_cloud" => "OLLAMA_API_KEY"
    }[provider]
    credential_ready = provider == "ollama" ||
      credential_present?(credential_name, File.join(config_dir, ".env"))
    executable_dir = File.join(Gem.user_dir, "bin")
    path_entries = ENV.fetch("PATH", "").split(File::PATH_SEPARATOR)

    io.puts "Boukensha doctor"
    io.puts "  version:           #{Boukensha::VERSION}"
    io.puts "  ruby:              #{RUBY_VERSION}"
    io.puts "  executable dir:    #{executable_dir}"
    io.puts "  executable on PATH: #{path_entries.include?(executable_dir) ? "yes" : "no"}"
    io.puts "  rc file:           #{rc_file}#{File.exist?(rc_file) ? "" : " (not found)"}"
    io.puts "  implementation:    #{implementation}"
    io.puts "  config dir:        #{config_dir}#{Dir.exist?(config_dir) ? "" : " (not found)"}"
    io.puts "  settings:          #{settings_file}#{File.exist?(settings_file) ? "" : " (not found)"}"
    io.puts "  provider:          #{provider || "(not configured)"}"
    io.puts "  model:             #{model || "(not configured)"}"
    io.puts "  credential ready:  #{credential_ready ? "yes" : "no"}"

    unless path_entries.include?(executable_dir)
      io.puts
      io.puts "Add the executable to this shell with:"
      io.puts %(  export PATH="#{executable_dir}:$PATH")
    end
  end

  def self.credential_present?(name, env_file)
    return true if name && ENV[name] && !ENV[name].strip.empty?
    return false unless name && File.exist?(env_file)

    File.foreach(env_file).any? do |line|
      key, value = line.strip.split("=", 2)
      key == name && value && !value.strip.delete_prefix('"').delete_suffix('"').empty?
    end
  end
  private_class_method :credential_present?
end

File: week1_baseline/python/12_context/boukensha_loader.py
"""Resolve a Boukensha implementation and report installation health."""

from __future__ import annotations

import importlib
import os
import sys
from pathlib import Path
from typing import Mapping, TextIO

import yaml


class LoaderError(Exception):
    """A user-facing configuration or implementation-selection error."""


def rc_file() -> Path:
    return Path.home() / ".boukensharc"


def load_rc(path: Path | None = None) -> dict[str, str]:
    path = Path(path) if path is not None else rc_file()
    if not path.exists():
        return {}

    try:
        parsed = yaml.safe_load(path.read_text(encoding="utf-8"))
    except yaml.YAMLError as error:
        raise LoaderError(f"invalid YAML in {path}: {error}") from error
    except OSError as error:
        raise LoaderError(f"cannot read {path}: {error}") from error

    if parsed is None:
        return {}
    if isinstance(parsed, str):
        return {"boukensha_path": parsed}
    if not isinstance(parsed, dict):
        raise LoaderError(f"{path} must contain a YAML mapping")
    return parsed


def _expand_rc_path(value, path: Path) -> Path | None:
    if not isinstance(value, str) or not value.strip():
        return None
    expanded = Path(os.path.expanduser(value))
    return expanded.resolve() if expanded.is_absolute() else (path.parent / expanded).resolve()


def resolve(
    *,
    path: Path | None = None,
    environ: Mapping[str, str] | None = None,
) -> tuple[Path | None, Path, dict[str, str]]:
    """Return (implementation override, config directory, parsed rc)."""
    path = Path(path) if path is not None else rc_file()
    environ = os.environ if environ is None else environ
    rc = load_rc(path)

    config = environ.get("BOUKENSHA_DIR")
    config_dir = (
        Path(os.path.expanduser(config)).resolve()
        if config is not None
        else _expand_rc_path(rc.get("boukensha_dir"), path) or Path.home() / ".boukensha"
    )

    source = environ.get("BOUKENSHA_PATH")
    implementation = (
        Path(os.path.expanduser(source)).resolve()
        if source is not None
        else _expand_rc_path(rc.get("boukensha_path"), path)
    )
    if implementation is not None and not (implementation / "boukensha" / "__init__.py").is_file():
        raise LoaderError(
            "no boukensha/__init__.py found at:\n"
            f"       {implementation}\n"
            f"       Check BOUKENSHA_PATH or {path}."
        )
    return implementation, config_dir.resolve(), rc


def load_implementation(
    *,
    path: Path | None = None,
    environ: dict[str, str] | None = None,
):
    environ = os.environ if environ is None else environ
    implementation, config_dir, _ = resolve(path=path, environ=environ)
    environ.setdefault("BOUKENSHA_DIR", str(config_dir))

    if implementation is None:
        return importlib.import_module("boukensha")

    source = str(implementation)
    if source not in sys.path:
        sys.path.insert(0, source)
    for name in tuple(sys.modules):
        if name == "boukensha" or name.startswith("boukensha."):
            del sys.modules[name]
    importlib.invalidate_caches()
    return importlib.import_module("boukensha")


def load_and_start_repl(
    *,
    path: Path | None = None,
    environ: dict[str, str] | None = None,
    output: TextIO | None = None,
    argv=None,
) -> None:
    environ = os.environ if environ is None else environ
    output = sys.stdout if output is None else output
    implementation, _, _ = resolve(path=path, environ=environ)
    if environ.get("BOUKENSHA_DEBUG"):
        print(f"[boukensha] loading from: {implementation or Path(__file__).parent}", file=output)

    package = load_implementation(path=path, environ=environ)
    repl = getattr(package, "repl", None)
    if not callable(repl):
        selected = implementation or Path(package.__file__).parent.parent
        raise LoaderError(
            f"the step at {selected} does not support the interactive REPL "
            "(added in step 8); run its examples directly or select step 8 or later"
        )
    arguments = list(sys.argv[1:] if argv is None else argv)
    no_tui = "--no-tui" in arguments
    repl(tui=not no_tui)


def _read_settings(settings_file: Path) -> dict:
    if not settings_file.exists():
        return {}
    try:
        parsed = yaml.safe_load(settings_file.read_text(encoding="utf-8"))
    except yaml.YAMLError as error:
        raise LoaderError(f"invalid YAML in {settings_file}: {error}") from error
    if parsed is None:
        return {}
    if not isinstance(parsed, dict):
        raise LoaderError(f"{settings_file} must contain a YAML mapping")
    return parsed


def _credential_present(name: str | None, env_file: Path, environ: Mapping[str, str]) -> bool:
    if name and environ.get(name, "").strip():
        return True
    if not name or not env_file.exists():
        return False
    try:
        lines = env_file.read_text(encoding="utf-8").splitlines()
    except OSError:
        return False
    for line in lines:
        key, separator, value = line.strip().partition("=")
        if separator and key == name and value.strip().strip("\"'"):
            return True
    return False


def doctor(
    *,
    path: Path | None = None,
    environ: Mapping[str, str] | None = None,
    output: TextIO | None = None,
    executable: str | None = None,
) -> None:
    from boukensha.version import VERSION

    environ = os.environ if environ is None else environ
    output = sys.stdout if output is None else output
    path = Path(path) if path is not None else rc_file()
    implementation, config_dir, _ = resolve(path=path, environ=environ)
    settings_file = config_dir / "settings.yaml"
    settings = _read_settings(settings_file)
    tasks = settings.get("tasks", {})
    player = tasks.get("player", {}) if isinstance(tasks, dict) else {}
    player = player if isinstance(player, dict) else {}
    provider = player.get("provider")
    model = player.get("model")
    credential_name = {
        "anthropic": "ANTHROPIC_API_KEY",
        "openai": "OPENAI_API_KEY",
        "gemini": "GEMINI_API_KEY",
        "ollama_cloud": "OLLAMA_API_KEY",
    }.get(provider)
    credential_ready = provider == "ollama" or _credential_present(
        credential_name, config_dir / ".env", environ
    )

    executable_path = Path(executable or sys.argv[0]).resolve()
    executable_dir = executable_path.parent
    path_entries = [Path(item).resolve() for item in environ.get("PATH", "").split(os.pathsep) if item]
    on_path = executable_dir in path_entries
    bundled = Path(__file__).resolve().parent

    print("Boukensha doctor", file=output)
    print(f"  version:            {VERSION}", file=output)
    print(f"  python:             {sys.version.split()[0]}", file=output)
    print(f"  executable:         {executable_path}", file=output)
    print(f"  executable dir:     {executable_dir}", file=output)
    print(f"  executable on PATH: {'yes' if on_path else 'no'}", file=output)
    print(f"  rc file:            {path}{'' if path.exists() else ' (not found)'}", file=output)
    print(f"  implementation:     {implementation or bundled}", file=output)
    print(f"  config dir:         {config_dir}{'' if config_dir.is_dir() else ' (not found)'}", file=output)
    print(f"  settings:           {settings_file}{'' if settings_file.exists() else ' (not found)'}", file=output)
    print(f"  provider:           {provider or '(not configured)'}", file=output)
    print(f"  model:              {model or '(not configured)'}", file=output)
    print(f"  credential ready:   {'yes' if credential_ready else 'no'}", file=output)
    if not on_path:
        print(f'\nAdd the executable directory to PATH:\n  export PATH="{executable_dir}:$PATH"', file=output)
